routes/course.js

```diff
@@ -63,57 +67,76 @@ router.get("/api/coursesByTerm/:plan_id", (req, res) => {
     .catch((err) => errorHandler(res, 400, err));
 });
 
-//add course, need to provide course info as json object in request body
-//distribution field is also updated
+// add course, need to provide course info as json object in request body
+// year, distribution, and fineReq objects updated
 router.post("/api/courses", async (req, res) => {
-  const course = req.body;
-  //console.log("course is ", course);
-  await plans
-    .findById(course.plan_id)
-    .then((plan) => {
-      //console.log(plan);
-      course.distribution_ids.forEach((id) => {
-        if (!plan.distribution_ids.includes(id)) {
-          errorHandler(res, 400, {
-            message: "Invalid combination of plan_id and distribution_ids.",
+  const courseBody = req.body;
+  await courses
+    .create(courseBody)
+    .then(async (retrievedCourse) => {
+      // find year obj and insert course id to array
+      await years
+        .findOneAndUpdate(
+          {
+            $and: [
+              { plan_id: retrievedCourse.plan_id },
+              { name: retrievedCourse.year },
+            ],
+          },
+          { $push: { courses: retrievedCourse._id } },
+          { new: true, runValidators: true }
+        )
+        .then(async (year) => {
+          retrievedCourse.year_id = year._id; // set year_id
+          await retrievedCourse.save();
+        });
+      // update plan's distribution objs
+      const plan = await plans.findById(retrievedCourse.plan_id);
+      for (let m_id of plan.major_ids) {
+        let distDoubleCount = undefined;
+        await distributions
+          .find({ plan_id: retrievedCourse.plan_id, major_id: m_id })
+          .then(async (distObjs) => {
+            for (let distObj of distObjs) {
+              if (
+                !distObjs.satisfied &&
+                (distDoubleCount === undefined ||
+                  distDoubleCount.length === 0 ||
+                  distDoubleCount.includes(distObj.name))
+              ) {
+                let isUpdated = await updateDistribution(
+                  distObj._id,
+                  retrievedCourse._id
+                );
+                if (isUpdated) {
+                  distDoubleCount = distObj.double_count;
+                }
+              }
+            }
           });
-        }
-      });
+      }
+      // return up to date course (because modified in helper method)
+
+      // option 1: return distribution and fine objs separate from course obj
+      // const updatedCourse = await courses.findById(retrievedCourse._id);
+      // const resp = { ...updatedCourse._doc, distributions: updatedDists, fineReqs: updatedFines };
+
+      // option 2: include distribution and fine objs within course obj
+      await courses
+        .findById(retrievedCourse._id)
+        .populate({
+          path: "distribution_ids",
+          populate: {
+            path: "fineReq_ids",
+          },
+        })
+        .then((course) => {
```

Comment:
Which is the smarter thing to do? 
What I realized is that when adding or deleting a course, the distribution / fineReq objects to return to frontend are all associated with the course object: course.distribution_ids and course.fineReq_ids 

So we have two options: 
1. return separately from course objs 
   - pros: readable, separate from course 
   - cons: have to build an array of dists and an array of fines 
2. return within course objs 
  - pros: compact, organized response body 
  - cons: not how the course obj works in frontend